packages/@lwc/engine-core/src/framework/renderer.ts

```diff
@@ -65,4 +67,14 @@ export interface RendererAPI {
         connectedCallback?: LifecycleCallback,
         disconnectedCallback?: LifecycleCallback
     ) => E;
+    registerContextProvider: (
+        element: E,
+        adapterContextToken: string,
+        onContextSubscription: WireContextSubscriptionCallback
+    ) => void;
+    registerContextConsumer: (
+        element: E,
+        adapterContextToken: string,
+        subscriptionPayload: WireContextSubscriptionPayload
+    ) => void;
```

Comment:
The changes in this module spoil its purity, somewhat, introducing non-DOM related things into the renderer interface. I anticipate this may be a controversial change, and I'm open to a better alternative.

Reply:
I'm not sure what else you could have done? If some implementation is different between `dom` and `server`, then we need to add a new API to `renderer.ts`.

If you're concerned that Locker doesn't care about this, then we could add these APIs to the list of "non-sandboxable" APIs: /cc @ravijayaramappa 

https://github.com/salesforce/lwc/blob/343a0785eb51458423096b3d0f0260b9f347e876/packages/@lwc/engine-dom/src/renderer-factory.ts#L21-L29

Reply:
I have to think about this. The renderer already give you the mechanisms (e.g.: addEventListener) to implement context, not sure why this has to be added there.